### ISSUE TITLE
Errors running convertfromhtml on Windows

### ISSUE THREAD

**Issue by Eiltherune** [basxsoftwareassociation/htmlgenerator#10]
I ran into a few hiccups while using this tool but I was able to fix them, so I figured you might want to implement the fixes in your code:

Issues:
1. The tool requires that black, bs4, and lxml be installed, but lxml is not easily available for Windows users and black and bs4 are flagged as "extra".
2. The tool uses the system's preferred encoding, "cp1252", when reading html files.

Solutions:
1. Use `os.name` to see what OS we're on. This adds an `if-else` statement to the `parsehtml` function:
   <details><summary>convertfromhtml.py</summary>
   <p>
  
   ```python
      def parsehtml(html, formatting, compact):
         out = [
             """import htmlgenerator as hg
      from htmlgenerator import mark_safe as s
      html = hg.BaseElement(""",
         ]
         if os.name == 'nt':
             parser = 'html.parser'
         else:
             parser = 'lxml'
         soup = BeautifulSoup(
             html,
             parser,
         )
         for subtag in soup.contents:
             tags = convert(subtag, 1, compact)
             if tags:
                 if tags[-1].strip():
                     tags[-1] += ","
                 out.extend(tags)
         out.append(")\n")
        
         separator = " " if compact else "\n"
         htmlstr = separator.join(filter(lambda line: bool(line.strip()), out))
         if not formatting:
             return htmlstr
         return black.format_file_contents(htmlstr, fast=True, mode=black.FileMode())
   ```
  
   </p>
   </details>

   Change `setup.py` to require lxml only if we're not on Windows, and to fully require black and bs4:
   <details><summary>setup.py</summary>
   <p>

   ```python
      setup(
          ...,
          install_requires=[
              "black", "beautifulsoup4", "lxml;platform_system!='Windows'",
          ],
      )
   ```
   </p>
   </details>
2. Add an additional optional flag to the command-line argument, allowing the user to specify an encoding. Default set to 'utf8':
   <details><summary>convertfromhtml.py</summary>
   <p>

   ```python
      def main():
          import sys
      
          formatflag = "--no-formatting"
          compactflag = "--compact"
          encodingflag = '--encoding'
      
          _encoding = 'utf8'
          files = sys.argv[1:]
          formatting = formatflag not in files
          compact = compactflag in files
          if formatflag in files:
              files.remove(formatflag)
          if compactflag in files:
              files.remove(compactflag)
          if encodingflag in files:
              _encoding = files[files.index(encodingflag) + 1]
              files.remove(encodingflag)
              files.remove(encoding)
          if not files:
              print(parsehtml(sys.stdin.read(), formatting, compact), end="")
          for _file in files:
              with open(_file, encoding=_encoding) as rf:
                  with open(_file + ".py", "w") as wf:
                      wf.write(parsehtml(rf.read(), formatting, compact))
   ```
   </p>
   </details>
   
Hopefully my edits will be of use to someone!

**Comment by saemideluxe** [comment]
Hi @Eiltherune 

Thanks for reporting these issues. I did not really expect anybody to use this on Windows,but glad if we can increase the compatability. Also, your provided code was very helpfull, I integrated your suggestions as described.

Just made a new release. 1.2.25 should contain all the changes. Let me know if this works for you as expected.

**Comment by Eiltherune** [comment]
I'm glad my edits were helpful! However, in `convertfromhtml.py` you left out the initialized value of `_encoding`, which is needed in case there is no `encodingflag` set. Just add the line `_encoding = 'utf8'` on line 153, and change line 169 to say `encoding=_encoding`, and everything should work properly.

Also, the `extras_require` in `setup.py` needs to either be changed to `install_require={"all":...` (as the dependencies are mandatory for all portions of htmlgenerator to function properly), or to `extras_require={"converter":...` to only install the dependencies if the user plans to use `convertfromhtml`. I believe the "proper" solution is to use `converter` because the core functionality of htmlgenerator does not need the dependencies. You should then add a note in `README.md` stating that the command-line tool needs to be installed by using `pip install "htmlgenerator[converter]"`, and mention the new encoding flag so that other Windows users don't need to hunt for it.

**Comment by saemideluxe** [comment]
Closing this, issue has been resolved in the comments of commit #0c0a8dcec50415a6ad43a07aa3330abccda0886b
